app-typescript/components/TreeMenu.tsx

```diff
@@ -0,0 +1,443 @@
+import * as React from "react";
+import { Icon } from "./Icon";
+import { createPopper, Instance } from '@popperjs/core';
+import {getPrefixedItemId, TreeSelectItem} from './TreeSelect/TreeSelectItem';
+import {keyboardNavigation} from './TreeSelect/KeyboardNavigation';
+import {createPortal} from 'react-dom';
+
+interface IState<T> {
+    options: Array<ITreeMenuNode<T>>;
+    openDropdown: boolean;
+    activeTree: Array<Array<ITreeMenuNode<T>>>;
+    buttonTree: Array<ITreeMenuNode<T>>;
+    buttonValue: ITreeMenuNode<T> | null;
+    filterArr: Array<ITreeMenuNode<T>>;
+    searchFieldValue: string;
+    firstBranchOptions: Array<ITreeMenuNode<T>>;
+
+    // array of classNames; used to focus an item after returning for another level
+    buttonTarget: Array<string>;
+}
+
+interface IProps<T> {
+    zIndex?: number;
+    searchPlaceholder?: string;
+    singleLevelSearch?: boolean;
+    getOptions?(): Array<ITreeMenuNode<T>>;
+    getLabel(item: T): string;
+    getId(item: T): string;
+    getBackgroundColor?(item: T): string;
+    getBorderColor?(item: T): string;
+    optionTemplate?(item: T): React.ComponentType<T> | JSX.Element;
+    children: (toggle: (event: React.SyntheticEvent) => void) => JSX.Element;
+}
+
+interface IParent<T> {
+    value: T;
+    children: Array<ITreeMenuNode<T>>;
+}
+
+interface IChildren<T> {
+    value: T;
+    onSelect(): void;
+}
+
+export type ITreeMenuNode<T> = IParent<T> | IChildren<T>;
+
+function nodeHasChildren<T>(item: IParent<T> | IChildren<T>): item is IParent<T> {
+    return item['children'] != null;
+}
+
+function nodeCanBeSelected<T>(item: IParent<T> | IChildren<T>): item is IChildren<T> {
+    return item['onSelect'] != null;
+}
+
+export class TreeMenu<T> extends React.Component<IProps<T>, IState<T>> {
```

Comment:
I see there's lot of code duplication between this and tree select. I expect we should now be able to reimplement tree select something like this:

```

export class TreeSelect extends React.Component {
    render() {
        return (
            <TreeMenu
                getOptions={() => options2}
                getId={(item) => item.name}
                getLabel={(item) => item.name}
                optionTemplate={(item) => {
                    return (
                        <div
                            style={{
                                display: 'flex',
                                alignItems: 'center',
                                gap: '4px'
                            }}
                        >
                            {item.icon}
                            {item.name}
                        </div>
                    )
                }}
            >
                {(toggle) => (
                    <div>
                        <div>label</div>

                        <div>
                            {
                                this.state.values.map((val, i) => (
                                    <span key={i}>{val}</span>
                                ))
                            }
                        </div>

                        <button onClick={toggle}>+</button>
                    </div>
                )}
            </TreeMenu>
        );
    }
}
```